### PR TITLE
[3.1] subjective-account-max-failures-window-size

### PR DESCRIPTION
**Pull request by huangminghuang**
This PR adds a new configuration option `--subjective-account-max-failures-window-size` to set the window size in number of blocks for `--subjective-account-max-failures`  option. 

Address IS #760

### CODE REVIEW

File: plugins/producer_plugin/producer_plugin.cpp [hunk 1]
```diff
@@ -673,10 +685,14 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
-                     FC_LOG_MESSAGE( error, "transaction ${id} exceeded failure limit for account ${a}",
-                                     ("id", trx->id())("a", first_auth) ) ) ) );
+            if (_account_fails.failure_limit(first_auth)) {
+               send_response(
+                     std::static_pointer_cast<fc::exception>(std::make_shared<tx_cpu_usage_exceeded>(FC_LOG_MESSAGE(
+                           error,
+                           "transaction ${id} exceeded failure limit for account ${a} until ${next_reset_timepoint}",
+                           ("id", trx->id())("a", first_auth)(
+                                 "next_reset_timepoint",
+                                 _account_fails.next_reset_timepoint(chain.head_block_num(), bt))))));
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:695]: Please cleanup this formatting.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 2]
```diff
@@ -2023,10 +2047,13 @@ bool producer_plugin_impl::process_unapplied_trxs( const fc::time_point& deadlin
...
-                  itr->next( std::make_shared<tx_cpu_usage_exceeded>(
-                        FC_LOG_MESSAGE( error, "transaction ${id} exceeded failure limit for account ${a}",
-                                        ("id", trx->id())("a", first_auth) ) ) );
+               if (itr->next) {
+                  itr->next(std::make_shared<tx_cpu_usage_exceeded>(FC_LOG_MESSAGE(
+                        error,
+                        "transaction ${id} exceeded failure limit for account ${a} until ${next_reset_timepoint}",
+                        ("id", trx->id())("a", first_auth)(
+                              "next_reset_timepoint",
+                              _account_fails.next_reset_timepoint(chain.head_block_num(), chain.head_block_time())))));
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:2056]: Please cleanup this formatting.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 3]
```diff
@@ -2439,8 +2466,7 @@ void producer_plugin_impl::produce_block() {
 
    block_state_ptr new_bs = chain.head_block_state();
 
-   _account_fails.report();
-   _account_fails.clear();
+   _account_fails.report(new_bs->block_num);
```
Comment (heifner) [inline plugins/producer_plugin/producer_plugin.cpp:2469]: I think you should make it `report_and_clear(uint32_t block_num)` and only call it from start_block. The `-1` in `report` vs `clear` seems a bit fragile.

**Review by heifner — CHANGES_REQUESTED** [review]

**Review by heifner — APPROVED** [review]

**Comment by heifner** [comment]
> Just curious, why do we do the `++` here?
> 
> https://github.com/AntelopeIO/leap/blob/71548a642577e0757f0e462bdaba20afa4f3841e/plugins/producer_plugin/producer_plugin.cpp#L223

Because we log the number of failures and want to include this failure in that number.

**Comment by greg7mdp** [comment]
> Because we log the number of failures and want to include this failure in that number.

But there is an add() function above. It seems that we should have an add() function incrementing the number of failures, and a `failure_limit` to only check if the limit was exceeded?

**Comment by heifner** [comment]
> > Because we log the number of failures and want to include this failure in that number.
> 
> But there is an add() function above. It seems that we should have an add() function incrementing the number of failures, and a `failure_limit` to only check if the limit was exceeded?

We don't want to add(), only increase the failure count.

**Review by greg7mdp — APPROVED** [review]

**Pull request merged**